File: TextClustering/src/com/yanjiuyanjiu/ml/kmeans/TextSegmentation.java
package com.yanjiuyanjiu.ml.kmeans;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * 英文分词.
 *
 * @author soulmachine@gmail.com
 *
 */
public final class TextSegmentation {
	/** 本类所有方法都是static，不需要实例化. */
	private TextSegmentation() {
		throw new AssertionError();
	}

	/**
	 * 将英文文本分词.
	 *
	 * @param text
	 *            文本文件
	 * @return 分割好的单词数组
	 */
	public static List<String> tokenizer(final String text) {
		final String lowerCaseText = text.toLowerCase(Locale.ENGLISH);
		final List<String> tokens = new ArrayList<String>();

		final StringBuilder sb = new StringBuilder();
		for (int i = 0; i < lowerCaseText.length(); i++) {
			final char character = lowerCaseText.charAt(i);
			if (Character.isLetter(character)) {
				sb.append(character);
			} else {
				if (sb.length() > 0) {
					tokens.add(sb.toString());
					sb.delete(0, sb.length());
				}
			}
		}
		if (sb.length() > 0) {
			tokens.add(sb.toString());
		}

		return ripStopWords(tokens);
	}

	/**
	 * 去掉停止词.
	 *
	 * @param oldText
	 *            分词后的单词数组
	 * @return 去掉停止词的新数组
	 */
	private static List<String> ripStopWords(final List<String> oldText) {
		// oldText - stopWords = temp
		final List<String> temp = new ArrayList<String>();
		for (final String word : oldText) {
			if (!StopWordsHandler.isStopWord(word)) { // 不是停用词
				temp.add(word);
			}
		}
		// String[] result = new String[temp.size()];
		// temp.toArray(result);
		// return result;
		return temp;
	}
}

File: TextClusteringScala/src/main/scala/com/yanjiuyanjiu/ml/kmeans/TextSegmentation.scala
/**
 * 分词.
 *
 * @author soulmachine@gmail.com
 * @date 2013-2-25
 * @version 0.1
 * @since Scala 2.10
 */
package com.yanjiuyanjiu.ml.kmeans

import scala.collection.mutable.ArrayBuffer

/**
 * 分词.
 *
 * @author soulmachine@gmail.com
 * @date 2013-2-25
 * @version
 * @since Scala 2.10
 */
object TextSegmentation {
  /**
   * 将英文文本分词.
   *
   * @param text
   *            文本文件
   * @return 分割好的单词数组
   */
  def tokenize(text: String): List[String] = {
    val lowerCaseText = text.toLowerCase()
    val tokens = new ArrayBuffer[String]

    val sb = new StringBuilder
    for (i <- 0 until lowerCaseText.length()) {
      val character = lowerCaseText.charAt(i)
      if (character.isLetter) {
        sb.append(character)
      } else {
        if (sb.length > 0) {
          tokens.append(sb.toString())
          sb.delete(0, sb.length())
        }
      }
    }
    if (sb.length > 0) {
      tokens.append(sb.toString())
    }

    ripStopWords(tokens)
  }
  /**
   * 去掉停止词.
   *
   * @param oldText
   *            分词后的单词数组
   * @return 去掉停止词的新数组
   */
  private def ripStopWords(tokens: ArrayBuffer[String]): List[String] = {
    val filtered = tokens.filter(!StopWordHandler.isStopWord(_))
    filtered.toList
  }
}